File: src/main/java/com/baulsupp/oksocial/secrets/Secrets.java
package com.baulsupp.oksocial.secrets;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.Writer;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Properties;
import java.util.Set;
import java.util.function.Function;

import static com.baulsupp.oksocial.util.Util.or;
import static com.google.common.collect.Sets.newHashSet;
import static java.util.Arrays.asList;
import static java.util.Optional.empty;
import static java.util.Optional.ofNullable;
import static java.util.stream.Collectors.joining;

public class Secrets {
  private static Secrets instance;

  private final Map<String, String> secrets;
  private final Function<String, Optional<String>> defaults;
  private final Optional<Path> file;
  private boolean changed = false;

  public Secrets(Map<String, String> secrets, Optional<Path> file,
      Function<String, Optional<String>> defaults) {
    this.secrets = secrets;
    this.defaults = defaults;
    this.file = file;
  }

  public Optional<String> get(String key) {
    Optional<String> result = ofNullable(secrets.get(key));

    if (!result.isPresent()) {
      result = defaults.apply(key);
    }

    return result.filter(s -> !s.isEmpty());
  }

  private void put(String key, String value) {
    secrets.put(key, value);
    changed = true;
  }

  public static Secrets loadSecrets() {
    Secrets classPathSecrets = loadClasspathDefaults();

    Path configFile =
        FileSystems.getDefault().getPath(System.getenv("HOME"), ".oksocial-secrets.properties");

    Properties p = new Properties();
    if (Files.exists(configFile)) {
      try (BufferedReader r = Files.newBufferedReader(configFile)) {
        p.load(r);
      } catch (IOException e) {
        e.printStackTrace();
      }
    }

    return new Secrets(new HashMap(p), Optional.of(configFile), classPathSecrets::get);
  }

  public static Secrets loadClasspathDefaults() {
    Properties p = new Properties();

    try (InputStream is = Secrets.class.getResourceAsStream("/oksocial-secrets.properties")) {
      if (is != null) {
        p.load(is);
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    return new Secrets(new HashMap(p), empty(), k -> empty());
  }

  public static Optional<String> getDefined(String key) {
    return instance().get(key);
  }

  public static String prompt(String name, String key, String defaultValue, boolean password) {
    Optional<String> defaulted = or(instance().getDefined(key), () -> ofNullable(defaultValue));

    String prompt = name + defaultDisplay(defaulted, password) + ": ";

    String value = "";

    if (System.console() != null) {
      if (password) {
        value = new String(System.console().readPassword(prompt));
      } else {
        value = System.console().readLine(prompt);
      }
    } else {
      System.err.println("using default value for " + key);
    }

    if (value.isEmpty()) {
      value = defaulted.orElse("");
    } else {
      instance().put(key, value);
    }

    return value;
  }

  public static Set<String> promptArray(String name, String key, Collection<String> defaults) {
    String valueString =
        prompt(name, key, defaults.stream().collect(joining(",")), false);
    return newHashSet(asList(valueString.split(",")));
  }

  private static String defaultDisplay(Optional<String> defaultValue, boolean password) {
    if (password) {
      defaultValue = defaultValue.map(s -> s.replaceAll(".", "\\*"));
    }

    return defaultValue.map(s -> " [" + s + "]").orElse("");
  }

  public static synchronized Secrets instance() {
    if (instance == null) {
      instance = loadSecrets();
    }

    return instance;
  }

  public void saveIfNeeded() throws IOException {
    if (changed && file.isPresent()) {
      Properties p = new Properties();
      p.putAll(secrets);

      try (Writer w = Files.newBufferedWriter(file.get())) {
        p.store(w, null);
      }
    }
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/secrets/Secrets.kt
package com.baulsupp.oksocial.secrets

import com.google.common.collect.Sets.newHashSet
import java.io.IOException
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import java.util.Arrays.asList

class Secrets(private val secrets: MutableMap<String, String>, private val file: Path?,
              private val defaults: (String) -> String?) {
    private var changed = false

    operator fun get(key: String): String? {
        var result = secrets[key] ?: defaults(key)

        return result?.let { if (it.isEmpty()) null else it }
    }

    private fun put(key: String, value: String) {
        secrets.put(key, value)
        changed = true
    }

    @Throws(IOException::class)
    fun saveIfNeeded() {
        if (changed && file != null) {
            val p = Properties()
            p.putAll(secrets)

            Files.newBufferedWriter(file).use { w -> p.store(w, null) }
        }
    }

    companion object {
        val instance: Secrets by lazy { loadSecrets() }

        fun loadSecrets(): Secrets {
            val classPathSecrets = loadClasspathDefaults()

            val configFile = FileSystems.getDefault().getPath(System.getenv("HOME"), ".oksocial-secrets.properties")

            val p = Properties()
            if (Files.exists(configFile)) {
                try {
                    Files.newBufferedReader(configFile).use { r -> p.load(r) }
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }

            return Secrets(p.toMutableStringMap(), configFile, { classPathSecrets[it] })
        }

        fun loadClasspathDefaults(): Secrets {
            val p = Properties()

            try {
                Secrets::class.java.getResourceAsStream("/oksocial-secrets.properties").use { s ->
                    if (s != null) {
                        p.load(s)
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }

            return Secrets(p.toMutableStringMap(), null, { _ -> null })
        }

        fun prompt(name: String, key: String, defaultValue: String, password: Boolean): String {
            val defaulted = instance[key] ?: defaultValue

            val prompt = name + defaultDisplay(defaulted, password) + ": "

            var value = ""

            if (System.console() != null) {
                if (password) {
                    value = String(System.console().readPassword(prompt))
                } else {
                    value = System.console().readLine(prompt)
                }
            } else {
                System.err.println("using default value for " + key)
            }

            if (value.isEmpty()) {
                value = defaulted
            } else {
                instance.put(key, value)
            }

            return value
        }

        fun promptArray(name: String, key: String, defaults: Iterable<String>): List<String> {
            val valueString = prompt(name, key, defaults.joinToString(","), false)
            return valueString.split("\\s*,\\s*".toRegex()).dropLastWhile { it.isEmpty() }
        }

        private fun defaultDisplay(defaultValue: String, password: Boolean): String {
            var display = if (password) {
                defaultValue.replace(".".toRegex(), "\\*")
            } else {
                defaultValue
            }

            return " [$display]"
        }
    }
}

private fun Properties.toMutableStringMap(): MutableMap<String, String> =
        this.entries.associate { it.key.toString() to it.value.toString() }.toMutableMap()
